File: sorting-searching/sorting-algorithms/replacement-selection-sort/replacement-selection-sort.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

vector<int> mergeTwoSortedRuns(const vector<int>& left, const vector<int>& right) {
    vector<int> result;

    int i = 0;
    int j = 0;

    while (i < left.size() && j < right.size()) {
        if (left[i] <= right[j]) {
            result.push_back(left[i]);
            i++;
        } else {
            result.push_back(right[j]);
            j++;
        }
    }

    while (i < left.size()) {
        result.push_back(left[i]);
        i++;
    }

    while (j < right.size()) {
        result.push_back(right[j]);
        j++;
    }

    return result;
}

vector<int> mergeAllRuns(vector<vector<int>>& runs) {
    if (runs.empty()) {
        return {};
    }

    vector<int> result = runs[0];

    for (int i = 1; i < runs.size(); i++) {
        result = mergeTwoSortedRuns(result, runs[i]);
    }

    return result;
}

vector<vector<int>> generateRuns(const vector<int>& input, int memorySize) {
    vector<vector<int>> runs;

    priority_queue<int, vector<int>, greater<int>> currentHeap;
    vector<int> frozen;

    int index = 0;
    int n = input.size();

    while (index < n && currentHeap.size() < memorySize) {
        currentHeap.push(input[index]);
        index++;
    }

    vector<int> currentRun;
    int lastOutput = -1000000000;

    while (!currentHeap.empty()) {
        int smallest = currentHeap.top();
        currentHeap.pop();

        currentRun.push_back(smallest);
        lastOutput = smallest;

        if (index < n) {
            int nextValue = input[index];
            index++;

            if (nextValue >= lastOutput) {
                currentHeap.push(nextValue);
            } else {
                frozen.push_back(nextValue);
            }
        }

        if (currentHeap.empty()) {
            runs.push_back(currentRun);
            currentRun.clear();

            for (int value : frozen) {
                currentHeap.push(value);
            }

            frozen.clear();
            lastOutput = -1000000000;
        }
    }

    return runs;
}

vector<int> replacementSelectionSort(const vector<int>& input, int memorySize) {
    vector<vector<int>> runs = generateRuns(input, memorySize);
    return mergeAllRuns(runs);
}

int main() {
    vector<int> arr = {6, 2, 8, 3, 9, 1, 7, 4, 5};
    int memorySize = 3;

    vector<vector<int>> runs = generateRuns(arr, memorySize);

    cout << "Generated Runs:\n";
    for (int i = 0; i < runs.size(); i++) {
        cout << "Run " << i + 1 << ": ";
        for (int value : runs[i]) {
            cout << value << " ";
        }
        cout << endl;
    }

    vector<int> sortedArray = mergeAllRuns(runs);

    cout << "\nFinal Sorted Output: ";
    for (int value : sortedArray) {
        cout << value << " ";
    }

    return 0;
}

File: sorting-searching/sorting-algorithms/replacement-selection-sort/replacement-selection-sort.js
class MinHeap {
  constructor() {
    this.heap = [];
  }

  size() {
    return this.heap.length;
  }

  isEmpty() {
    return this.heap.length === 0;
  }

  push(value) {
    this.heap.push(value);
    this.heapifyUp();
  }

  pop() {
    if (this.isEmpty()) {
      return null;
    }

    if (this.heap.length === 1) {
      return this.heap.pop();
    }

    const root = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.heapifyDown();

    return root;
  }

  heapifyUp() {
    let index = this.heap.length - 1;

    while (index > 0) {
      const parent = Math.floor((index - 1) / 2);

      if (this.heap[parent] <= this.heap[index]) {
        break;
      }

      [this.heap[parent], this.heap[index]] = [this.heap[index], this.heap[parent]];
      index = parent;
    }
  }

  heapifyDown() {
    let index = 0;

    while (true) {
      let smallest = index;
      const left = 2 * index + 1;
      const right = 2 * index + 2;

      if (left < this.heap.length && this.heap[left] < this.heap[smallest]) {
        smallest = left;
      }

      if (right < this.heap.length && this.heap[right] < this.heap[smallest]) {
        smallest = right;
      }

      if (smallest === index) {
        break;
      }

      [this.heap[index], this.heap[smallest]] = [this.heap[smallest], this.heap[index]];
      index = smallest;
    }
  }
}

function mergeTwoSortedRuns(left, right) {
  const result = [];

  let i = 0;
  let j = 0;

  while (i < left.length && j < right.length) {
    if (left[i] <= right[j]) {
      result.push(left[i]);
      i++;
    } else {
      result.push(right[j]);
      j++;
    }
  }

  while (i < left.length) {
    result.push(left[i]);
    i++;
  }

  while (j < right.length) {
    result.push(right[j]);
    j++;
  }

  return result;
}

function mergeAllRuns(runs) {
  if (runs.length === 0) {
    return [];
  }

  let result = runs[0];

  for (let i = 1; i < runs.length; i++) {
    result = mergeTwoSortedRuns(result, runs[i]);
  }

  return result;
}

function generateRuns(inputArray, memorySize) {
  const runs = [];
  const currentHeap = new MinHeap();
  let frozen = [];

  let index = 0;

  while (index < inputArray.length && currentHeap.size() < memorySize) {
    currentHeap.push(inputArray[index]);
    index++;
  }

  let currentRun = [];
  let lastOutput = -Infinity;

  while (!currentHeap.isEmpty()) {
    const smallest = currentHeap.pop();

    currentRun.push(smallest);
    lastOutput = smallest;

    if (index < inputArray.length) {
      const nextValue = inputArray[index];
      index++;

      if (nextValue >= lastOutput) {
        currentHeap.push(nextValue);
      } else {
        frozen.push(nextValue);
      }
    }

    if (currentHeap.isEmpty()) {
      runs.push(currentRun);
      currentRun = [];

      for (const value of frozen) {
        currentHeap.push(value);
      }

      frozen = [];
      lastOutput = -Infinity;
    }
  }

  return runs;
}

function replacementSelectionSort(inputArray, memorySize) {
  const runs = generateRuns(inputArray, memorySize);
  return mergeAllRuns(runs);
}

const arr = [6, 2, 8, 3, 9, 1, 7, 4, 5];
const memorySize = 3;

const runs = generateRuns(arr, memorySize);

console.log("Generated Runs:");
runs.forEach((run, index) => {
  console.log(`Run ${index + 1}:`, run);
});

const sortedArray = mergeAllRuns(runs);

console.log("Final Sorted Output:", sortedArray);
